gnome-initial-setup/demo/eos-demo-video.c

```diff
@@ -0,0 +1,230 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */
+/*
+ * Copyright (C) 2017 Endless Mobile, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *   Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include "config.h"
+
+#include <eos-demo-video.h>
+
+typedef struct _GisDemoVideoPrivate
+{
+  GtkWindow *main_window;
+  GstElement *element;
+  GstElement *player;
+  GtkWidget *widget;
+  GstBus *bus;
+} GisDemoVideoPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (GisDemoVideo, gis_demo_video, GTK_TYPE_APPLICATION)
+
+static void
+on_demo_video_event (GtkWidget *widget,
+                     GdkEvent  *event,
+                     gpointer  user_data)
+{
+  if (event->type == GDK_BUTTON_PRESS ||
+      event->type == GDK_2BUTTON_PRESS ||
+      event->type == GDK_3BUTTON_PRESS ||
+      event->type == GDK_KEY_PRESS ||
+      event->type == GDK_MOTION_NOTIFY)
+    {
+      gtk_widget_destroy (widget);
+    }
+}
+
+static gboolean
+on_video_bus_event (GstBus     *bus,
+                    GstMessage *msg,
+                    gpointer   user_data)
+{
+  GisDemoVideo *demo_video = GIS_DEMO_VIDEO (user_data);
+  GisDemoVideoPrivate *priv = gis_demo_video_get_instance_private (demo_video);
+
+  switch (GST_MESSAGE_TYPE (msg))
+    {
+    case GST_MESSAGE_EOS:
+      gst_element_set_state (priv->player, GST_STATE_PAUSED);
+      gst_element_seek_simple (priv->player,
+                               GST_FORMAT_TIME,
+                               GST_SEEK_FLAG_FLUSH |
+                               GST_SEEK_FLAG_ACCURATE,
+                               0);
+      gst_element_set_state (priv->player, GST_STATE_PLAYING);
+    default:
+      break;
+    }
+
+  return G_SOURCE_CONTINUE;
+}
+
+#define DEMO_VIDEO_FILE_URI "file://" DEMOVIDEODATADIR "/eos-demo-video.ogv"
+
+/* The shell hides all windows on startup, by showing
+ * the overview which is understandable, so show the
+ * window a few seconds later */
```

Comment:
The timeout makes a difference because we show the window *after* the icon grid has been shown.

But in any case lets patch (1) so that we can drop this workaround.